Add new utility `create_date_files`

rnyb opened the pull request:
Closes #828

rnyb committed:
Add script to generate date files from yaml config

- Add create_date_files.py to generate single_dates.txt and diff_dates.txt
- Parse global variable YAML for date lists and validate their format
- Support command-line arguments for input file and date list keys
- Output files compatible with ECLRST2ROFF and ECLDIFF2ROFF

rnyb committed:
test(create_date_files): add tests for date and config validation

- Add unit tests for is_iso_date_item with date, datetime, and string inputs
- Add tests for is_iso_date_list and is_iso_diffdate_list with valid and invalid cases
- Add tests for validate_cfg covering success, optional None, and missing keys

rnyb committed:
Fix lint issues.

rnyb committed:
feat(create_date_files): add package init file

rnyb committed:
fix import to use subscript.create_date_files in test file

rnyb committed:
Add documentation for create_date_files script

rnyb committed:
add CREATE_DATE_FILES step and CLI entry point

- Register create_date_files as a CLI tool in pyproject.toml
- Add CreateDateFiles forward model step with documentation and example
- Import create_date_files in forward_model_steps for plugin registration
- Fix minor formatting in create_date_files.py description string

rnyb committed:
create_date_files: add tests for main function and refactor imports

- Add test_main to verify CLI execution and output file creation
- Refactor imports to use create_date_files directly instead of alias
- Improve test formatting and coverage for date validation functions

rnyb committed:
test(create_date_files): make main test use tmp_path and validate outputs

- import textwrap and use textwrap.dedent to write the YAML fixture
- remove unused tmp_path parameter from helper test
- monkeypatch.chdir(tmp_path) to create hardcoded output files in tmp dir
- pass variable key names using --single-dates and --diff-dates flags
- assert generated files exist in tmp_path and verify their contents

rnyb committed:
create_date_files fixes

larsevj commented on src/subscript/create_date_files/create_date_files.py:
```diff
@@ -0,0 +1,185 @@
+    return True
+
+
+def validate_cfg(cfg, single_dates: str | None, diff_dates: str | None) -> bool:
```
Should there be some early check if both single_dates and diff_dates is None, and maybe a warning?

rnyb committed:
Change cfg_global variable name to cfg_dates

rnyb committed:
fix(create_date_files): use valid isinstance checks

- Replace isinstance(item, datetime.date | datetime.datetime) with
- isinstance(item, datetime.date) since datetime.datetime is a subclass
- Change isinstance(..., list | tuple) to isinstance(..., (list, tuple))
- Use (list, tuple) for pair checks to validate diffdate pairs
- Prevent TypeError from using PEP 604 unions in isinstance and improve
- compatibility across Python versions

rnyb committed:
Add utf-8 encoding to output files

rnyb committed:
fix(create_date_files): warn when date lists are empty

rnyb committed:
Check that not both single-dates and diff-dates are None

rnyb committed:
refactor(create_date_files): add typing and improve config validation

- Add typing and annotations for parser, validation, and main
- Strengthen validate_cfg to check None, dict type and sections
- Fix isinstance checks for diffdate pairs and assert for mypy
- Simplify f-strings and update globvar_file help text to mention YAML

rnyb commented on src/subscript/create_date_files/create_date_files.py:
```diff
@@ -0,0 +1,185 @@
+    cfg_global = cfg["global"]["dates"]
+
+    if single_dates is not None:
+        if single_dates not in cfg_global:
```
I couldn't figure out how to do it differently.

larsevj replied:
I think I meant more that you can change this `_validate_*` functions into one, and then pass the content validation function as an argument, for instance `_validate_dates_object(single_dates, validation_func=is_iso_date_list)` or something similar?

rnyb replied:
Maybe we can keep them separate for clarity even if it is less compact?

rnyb committed:
refactor(create_date_files): tighten type for is_iso_diffdate_list

- Use list typing for diffdate_list parameter
- Remove redundant isinstance check for list/tuple
- Keep validation that each element is a length-2 sequence of ISO dates

rnyb commented on src/subscript/create_date_files/create_date_files.py:
```diff
@@ -0,0 +1,185 @@
+    return all(is_iso_date_item(it) for it in date_list)
+
+
+def is_iso_diffdate_list(diffdate_list) -> bool:
```
Good point. Removed redundant check.

rnyb committed:
docs(forward_model_steps): correct example parameter name GLOBVARFILE

- Fix typo in example usage: change <GLOBVAR_FILE> to <GLOBVARFILE>
- File: src/subscript/hook_implementations/forward_model_steps.py

rnyb committed:
refactor(create_date_files): split validate_cfg into focused helpers

- Extract _validate_cfg_structure to validate basic config structure
- Add _validate_single_dates and _validate_diff_dates for specific checks
- Simplify validate_cfg to orchestrate helpers and return results
- Improve and clarify logger messages for empty or invalid config
- Preserve existing warnings for non-ISO date formats

rnyb commented on src/subscript/create_date_files/create_date_files.py:
```diff
@@ -0,0 +1,219 @@
+    return all(is_iso_date_item(it) for it in date_list)
+
+
+def is_iso_diffdate_list(diffdate_list: list) -> bool:
```
See https://github.com/equinor/subscript/pull/826#discussion_r2455254163
@larsevj keep or change back?

rnyb committed:
refactor: inline date list validation and remove unused helpers

- Remove is_iso_date_list and is_iso_diffdate_list helper functions
- Inline date list validation logic directly in _validate_single_dates and
  _validate_diff_dates for clarity
- Add improved warnings for invalid diff date formats

rnyb committed:
Remove redundant tests for is_iso_date_list and is_iso_diffdate_list

Delete tests for deleted functions

rnyb committed:
feat(create-date-files): validate mapping args and executable

- Introduce ClassVar placeholders to avoid magic strings
- Resolve executable path once and verify presence
- Add validate_pre_experiment method checking mapping and args
- Raise ForwardModelStepValidationError with aggregated messages
- Refactor command construction using constant placeholders

rnyb commented on src/subscript/hook_implementations/forward_model_steps.py:
```diff
@@ -92,6 +93,35 @@ def documentation() -> ForwardModelStepDocumentation | None:
+                "--diff-dates",
+                "<DIFFDATES>",
+            ],
+        )
```
For attempt at pre validation, see https://github.com/equinor/subscript/pull/826/commits/f5e3ae17f3b0d47cb6d091f41ea24fccd173528b

rnyb committed:
fix(create-date-files): validate diff date entries are lists

rnyb commented on src/subscript/create_date_files/create_date_files.py:
```diff
@@ -0,0 +1,236 @@
+    if not cfg_dates[diff_dates]:
+        logger.warning(f"{diff_dates} is empty")
+
+    for pair in cfg_dates[diff_dates]:
```
Updated

rnyb commented on src/subscript/hook_implementations/forward_model_steps.py:
```diff
@@ -92,6 +96,77 @@ def documentation() -> ForwardModelStepDocumentation | None:
+        if self.command[0] is None:
+            errors.append("Executable 'create_date_files' not found in PATH.")
+
+        mapping = fm_step_json.get("mapping", {})
```
Ok, I will remove that part of the code

rnyb committed:
Rewrite fm pre-validation

rnyb committed:
Add sys.exit if validation error.

rnyb committed:
Fix import

larsevj commented on src/subscript/create_date_files/create_date_files.py:
```diff
@@ -0,0 +1,240 @@
+
+    for pair in cfg_dates[diff_dates]:
+        if not isinstance(pair, list):
+            logger.warning("Each diff date entry must be a list of two dates.")
```
Should these be `logger.errors` since they exits the script?

rnyb replied:
see commit https://github.com/equinor/subscript/pull/826/commits/804235e8f5bd213bd5a0084d5646203704ed26b4

rnyb committed:
Replace warnings with errors for invalid date format and structure

larsevj reviewed (approved):
LGTM, great work 🚀 Just remember to squash the commits!
